File: docs/examples/todo/node/src/services/TodoService.js
const TodoRepository = require('../repositories/TodoRepository');

/**
 * Business-logic layer for todo operations.
 *
 * Depends on the {@link TodoRepository} abstraction (Dependency Inversion)
 * and owns all validation and domain rules (Single Responsibility).
 */
class TodoService {
  /**
   * @param {TodoRepository} repository - Concrete repository implementation.
   */
  constructor(repository) {
    this._repository = repository;
  }

  /**
   * Create a new todo after validating inputs.
   *
   * @param {string} title - Required, max 200 characters.
   * @param {string} [description=''] - Optional description.
   * @returns {Promise<{ id: string, title: string, description: string, done: boolean, createdAt: string }>}
   * @throws {Error} If title is empty or exceeds 200 characters.
   */
  async create(title, description = '') {
    if (!title || title.trim().length === 0) {
      throw new Error('Title is required');
    }
    if (title.length > 200) {
      throw new Error('Title must not exceed 200 characters');
    }
    return this._repository.save({ title: title.trim(), description });
  }

  /**
   * List all todos, newest first.
   *
   * @returns {Promise<Array<{ id: string, title: string, description: string, done: boolean, createdAt: string }>>}
   */
  async list() {
    const todos = await this._repository.findAll();
    return todos.sort((a, b) => b.createdAt.localeCompare(a.createdAt));
  }

  /**
   * Mark an existing todo as done.
   *
   * @param {string} id - Todo identifier.
   * @returns {Promise<{ id: string, title: string, description: string, done: boolean, createdAt: string }>}
   * @throws {Error} If the todo is not found.
   */
  async complete(id) {
    const todo = await this._repository.markDone(id);
    if (!todo) {
      throw new Error('Todo not found');
    }
    return todo;
  }
}

module.exports = TodoService;

File: docs/examples/todo/python/app/services/todo_service.py
"""Business-logic layer for todo operations (SRP + DIP)."""

from app.repositories.todo_repository import TodoRepository


class TodoService:
    """Owns all validation and domain rules for todos.

    Depends on the TodoRepository abstraction, not on a concrete store.
    """

    def __init__(self, repository: TodoRepository) -> None:
        """Initialise the service with a repository implementation.

        Args:
            repository: Concrete TodoRepository.
        """
        self._repository = repository

    def create(self, title: str, description: str = "") -> dict:
        """Create a new todo after validating inputs.

        Args:
            title: Required, max 200 characters.
            description: Optional description text.

        Returns:
            Created todo dict.

        Raises:
            ValueError: If title is empty or exceeds 200 characters.
        """
        if not title or not title.strip():
            raise ValueError("Title is required")
        if len(title) > 200:
            raise ValueError("Title must not exceed 200 characters")
        return self._repository.save(title.strip(), description)

    def list_all(self) -> list[dict]:
        """List all todos, newest first.

        Returns:
            Sorted list of todo dicts.
        """
        todos = self._repository.find_all()
        return sorted(todos, key=lambda t: t["created_at"], reverse=True)

    def complete(self, todo_id: str) -> dict:
        """Mark an existing todo as done.

        Args:
            todo_id: Unique todo identifier.

        Returns:
            Updated todo dict.

        Raises:
            ValueError: If the todo is not found.
        """
        todo = self._repository.mark_done(todo_id)
        if todo is None:
            raise ValueError("Todo not found")
        return todo
